finality-aleph/src/justification/requester.rs

```diff
@@ -0,0 +1,150 @@
+{
+    pub fn new(
+        block_requester: RB,
+        client: Arc<C>,
+        finalizer: F,
+        justification_request_delay: D,
+        metrics: Option<Metrics<<B::Header as Header>::Hash>>,
+    ) -> Self {
+        BlockRequester {
+            attempt: 0,
+            last_seen_finalized: 0u32.into(),
+            block_requester,
+            client,
+            finalizer,
+            justification_request_delay,
+            metrics,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn handle_justification_notification(
+        &mut self,
+        notification: JustificationNotification<B>,
+        verifier: V,
+        last_finalized: NumberFor<B>,
+        stop_h: NumberFor<B>,
+    ) {
+        let JustificationNotification {
+            justification,
+            number,
+            hash,
+        } = notification;
+
+        if number <= last_finalized || number > stop_h {
+            debug!(target: "afa", "Not finalizing block {:?}. Last finalized {:?}, stop_h {:?}", number, last_finalized, stop_h);
+            return;
+        };
+
+        if !(verifier.verify(&justification, hash)) {
+            warn!(target: "afa", "Error when verifying justification for block {:?} {:?}", number, hash);
+            return;
+        };
+
+        debug!(target: "afa", "Finalizing block {:?} {:?}", number, hash);
+        let finalization_res = self.finalizer.finalize_block(
+            hash,
+            number,
+            Some((ALEPH_ENGINE_ID, justification.encode())),
+        );
+        match finalization_res {
+            Ok(()) => {
+                self.justification_request_delay.on_block_finalized();
+                debug!(target: "afa", "Successfully finalized {:?}", number);
+                if let Some(metrics) = &self.metrics {
+                    metrics.report_block(hash, Instant::now(), Checkpoint::Finalized);
+                }
+            }
+            Err(e) => {
+                error!(target: "afa", "Fail in finalization of {:?} {:?} -- {:?}", number, hash, e);
+            }
+        }
+    }
+
+    fn maintain_requests(&mut self) {
+        if self.client.info().finalized_number == self.last_seen_finalized {
+            self.attempt += 1;
+        } else {
+            self.attempt = 0;
+            self.last_seen_finalized = self.client.info().finalized_number;
+        }
+
+        if self.attempt == MAX_ATTEMPS {
+            self.block_requester.clear_justification_requests();
+            self.attempt = 0;
+        }
+    }
+
+    pub fn request_justification(&mut self, num: NumberFor<B>) -> NumberFor<B> {
+        self.maintain_requests();
+
+        let num = if num > self.client.info().best_number
+            && self.client.info().best_number > MIN_ALLOWED_DELAY.into()
+        {
+            self.client.info().best_number - MIN_ALLOWED_DELAY.into()
+        } else {
+            num
+        };
+
+        if self.justification_request_delay.can_request_now() {
```

Comment:
The comment higher up in this file sorta overrides this problem but worth mentioning – note that by calling `maintain_requests` outside of this `if` you are increasing the attempt counter even without attempts.

Reply:
Done